Fix error enumerablility

BridgeAR opened the pull request:
1. Commit:

A error message should always be non-enumerable. This makes sure
that is true for dns errors as well. It also adds another check
in `common.expectsError` to make sure no other regressions are
introduced going forward.

Fixes: #19716

<s>2. Commit:

Right now the hostname could in some cases be missed, depending on
the libuv error number. This makes sure there the hostname is always
added, if available.</s>

__Update__: I removed the second commit to another PR so it is easier for backporting.

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/doc/guides/contributing/pull-requests.md#commit-message-guidelines)

BridgeAR commented:
CI https://ci.nodejs.org/job/node-test-pull-request/13967/

joyeecheung commented:
BTW I think I previously kept the hostname from the error messages in the case of libuv errors because it **changes the error messages**, which is semver-major. If we keep the changes, do we need to make this semver-major? Or should we backport that in a semver-patch manner?

BridgeAR commented:
@joyeecheung I was wondering about the semverness of this as well. I personally would argue this is a patch because the hostname was applied in case the libuv error code would have been one of the special handled ones (that are relatively common) but not for any other ones. Of course also a bugfix can be semver-major, so I would like to defer the decision about the semverness to others. If we declare it as being semver-major, I am also happy to backport the first commit only.

BridgeAR commented:
I added another commit to always add the hostname, if available as suggested by @joyeecheung.

New CI https://ci.nodejs.org/job/node-test-pull-request/13970/

alexjeffburke commented:
I think it would be better to avoid the hostname if it's going to make this change server major. Bear in mind, this behavioural change was originally introduced in a minor version (9.7.0) so I think it would be beneficial to patch just that different as a minor and move the hostname thing into a separate patch.

BridgeAR commented:
I updated the PR to only fix the bug and moved the second commit to another PR so it is easier while backporting.

CI https://ci.nodejs.org/job/node-test-pull-request/13993/

BridgeAR commented:
Rebased due to conflicts. CI before landing: https://ci.nodejs.org/job/node-test-pull-request/14039/

BridgeAR commented:
Landed in 22da2f731d90235289a95bffe1ae3750a62a33bc

alexjeffburke commented:
Awesome that it landed! Thanks all.

targos commented:
This needs a backport PR for v9.x-staging.
